sable_history/src/pg_history_service.rs

```diff
@@ -0,0 +1,176 @@
+            // TODO: PMs
+            return Err(HistoryError::InvalidTarget(target));
+        };
+
+        let mut connection_lock = self.database_connection.lock().await;
+
+        let db_channel_id = i64::try_from(channel_id.as_u64()).expect("channel id overflows u64");
+        if channels::dsl::channels
+            .find(db_channel_id)
+            .select(crate::models::Channel::as_select())
+            .first(&mut *connection_lock)
+            .await
+            .optional()
+            .expect("Could not check if channel exists")
+            .is_none()
+        {
+            return Err(HistoryError::InvalidTarget(target));
+        }
+
+        match request {
+            HistoryRequest::Latest { to_ts, limit } => {
+                let limit = i64::min(10000, i64::try_from(limit).unwrap_or(i64::MAX));
+                Ok(match to_ts {
+                    Some(to_ts) => {
+                        // Lowest UUIDv7 corresponding to the timestamp
+                        let to_uuid = uuid::Builder::from_unix_timestamp_millis(
+                            u64::try_from(to_ts)
+                                .unwrap_or(u64::MIN) // floor timestamps to Epoch
+                                .saturating_mul(1000),
+                            &[u8::MIN; 10],
+                        )
+                        .into_uuid();
+                        Box::new(
+                            messages::dsl::messages
+                                .filter(messages::dsl::target_channel.eq(db_channel_id))
+                                .filter(messages::dsl::id.lt(to_uuid))
+                                .order(messages::dsl::id.desc())
+                                .limit(limit)
+                                .load_stream(&mut *connection_lock),
+                        )
+                    }
+                    None => Box::new(
+                        messages::dsl::messages
+                            .filter(messages::dsl::target_channel.eq(db_channel_id))
+                            .order(messages::dsl::id.desc())
+                            .limit(limit)
+                            .load_stream(&mut *connection_lock),
+                    ),
+                }
+                .await
+                .expect("could not query messages")
+                .map(|row| {
+                    let crate::models::Message {
+                        id,
+                        source_user,
+                        target_channel,
+                        text,
+                    } = row?;
+                    Ok(HistoryLogEntry {
+                        id: (),
+                        details: NetworkStateChange::NewMessage(update::NewMessage {
+                            message: (),
+                            source: (),
+                            target: (),
+                        }),
+                        source_event: (),
+                        timestamp: (),
+                    })
```

Comment:
@spb Could you fill the blanks here?

Reply:
I don't think they can be filled in as-is - the IDs might not be meaningful to the requesting server if those messages have aged out of local retention. I think the interface needs modifying to return the information directly instead of by ID reference.

Reply:
I can work on those later after I know exactly which ones need to be removed

Reply:
OK, so in order of the above code:

- HistoryLogEntry ID isn't needed
- Message content, source, and target will be required, but will need to have the actual content and not IDs
- Source event ID probably isn't useful. We might need some information from the event to fill in capability-based tags, but the event itself will be long gone so those fields will need to be stored in the history db as their own separate fields
- Timestamp will be needed for server_time at least

...but honestly I think starting from here and trying to remove things is the wrong way to think about it. We know what information the chathistory handler needs to be able to send the messages to the client, so the better answer is to just write a struct that contains that information in the most logical manner, then make sure that the database contains that information and that the local history service can translate from the in-memory log into that format.

Reply:
"We know what information the chathistory handler needs to be able to send the messages to the client" -> currently, the chathistory handler reuses existing code through `ClientServer::send_item`; adding a new struct means we need to either duplicate or rewrite everything in `sable_ircd/src/messages/send_history.rs`

Reply:
Ah, or we could directly build the same `OutboundClientMessage` that `ClientServer::send_item` would before sending, hmm

Reply:
Right. The new chathistory command handler won't be able to rely on `send_item` because it can't use a `HistoryLogEntry`, because the historic user state that it would need to refer to doesn't exist any more. To support remote history it'll need to construct the outbound client message directly, including getting the tag data from the history message provided by the history server and using string data instead of user/channel objects for the source and target.